### ISSUE TITLE
Kerner parameter "vfio-pci.ids" is ignored.

### ISSUE THREAD

**Issue by dron1885** [anatol/booster#20]
With vfio-pci.ids=1002:67df,1002:aaf0 secondary GPU should not be activated.

**Comment by anatol** [comment]
I am quite understand the nature of this problem. Could you please provide more information what exactly the issue is, how to reproduce it and why/how this parameter should be handled by booster?

**Comment by GabTheBab** [comment]
I believe I have the same problem, so hopefully I can describe it better. The kernel command line option vfio-pci.ids doesnt seem to work with booster images, while the same command line option works with a dracut image. Normally it would assign the vfio-pci driver to the devices you specify, and thats how the same boot entry works when pointed to a dracut image.

**Comment by anatol** [comment]
Booster does not do anythong with this parameter. It just passed over to linux kernel.

I am looking at dracut project and do not see any mention of it https://github.com/dracutdevs/dracut/search?q=vfio-pci.ids
neither systemd contains any references https://github.com/systemd/systemd/search?q=vfio-pci.ids

I only see it in the linux kernel here https://github.com/torvalds/linux/blob/719bbd4a509f403f537adcaefd8ce17532be2e84/drivers/vfio/pci/vfio_pci.c#L39

So the loadable kernel module `vfio_pci` is the one that handles this parameter. You need to make sure the module is added to the image. Add following line to the booster config `modules: vfio_pci`. Additionally you may want to enable the booster debug and make sure the module is loaded at the boot time.

**Comment by GabTheBab** [comment]
modules: vfio_pci,vfio,vfio_iommu_type1,vfio_virqfd

this is whats inside my booster.yaml, i see the modules inside of the initramfs if i use lsinitrd, running booster debug, i can't seem to find where its logging stuff to, I dont see anything about it inside of journalctl for that boot.

**Comment by GabTheBab** [comment]
https://wiki.archlinux.org/index.php/PCI_passthrough_via_OVMF#Loading_vfio-pci_early 

Here is how things would normally be done on mkinitcpio/dracut for reference

**Comment by anatol** [comment]
@GabTheBab to enable log add `booster.debug` kernel parameter to your bootloader config. Once it is rebooted you'll see a lot of text during the boot process.

thanks for the link. The equivalent of if for booster would be the modules like you specified.

The only question I have here whether the kernel autoload it (using modalias) or booster has to force these modules loading. The force loading does not work yet but I can come up with a quick proof-of-concenp if you are willing to test it.

**Comment by GabTheBab** [comment]
I do have booster.debug=1 set in my systemd boot entry, and I'm definitely willing to do some testing if needed. The text during the boot process sadly doesn't stay on screen long enough for me to read it, is there any place i could see the debug info after boot?

**Comment by anatol** [comment]
> is there any place i could see the debug info after boot?

It currently not, but I filed #60 to investigate it

**Comment by anatol** [comment]
Please try this branch https://github.com/anatol/booster/tree/force_modules_loading it adds a new booster.yaml config option `modules_force_load: ` that specifies what modules need to be explicitly loaded at the boot time. Add your vfio modules there.

**Comment by GabTheBab** [comment]
Built commit c1a3246, added the new setting to booster.yaml and regenerated the initramfs, sadly still seems to load the nvidia driver rather than vfio-pci.

**Comment by GabTheBab** [comment]
modules: vfio_pci,vfio,vfio_iommu_type1,vfio_virqfd
modules_force_load: vfio_pci,vfio,vfio_iommu_type1,vfio_virqfd
compression: zstd

my full booster.yaml

**Comment by anatol** [comment]
It turns out that there was a circular dependencies between `vfio` and `vfio_iommu_type1`

```
modules.softdep:softdep vfio post: vfio_iommu_type1 vfio_iommu_spapr_tce
modules.dep:kernel/drivers/vfio/vfio_iommu_type1.ko.xz: kernel/drivers/vfio/vfio.ko.xz
```
that booster did not handle correctly. I added a quick fix for this issue and now I see booster loads these modules at the boot time.

I pushed it to `force_modules_loading`. Please try it again.


WRT your config `compression: zstd` is the default option, no need to specify it. `modules_force_load` implicitly added to `modules` list, no need to specify it twice. Following config is equivalent to one printed above.

```
modules_force_load: vfio_pci,vfio,vfio_iommu_type1,vfio_virqfd
```

**Comment by GabTheBab** [comment]
force_modules_loading or force_modules_load?

**Comment by anatol** [comment]
Sorry for the confusion it is `modules_force_load` branch.

**Comment by GabTheBab** [comment]
Built d620be6, config is just `modules_force_load: vfio_pci,vfio,vfio_iommu_type1,vfio_virqfd`, rebuilt initramfs, still loading nvidia driver.

**Comment by anatol** [comment]
Once you booted what is the output of your `lsmod | grep vfio` and `cat /proc/cmdline`?

**Comment by GabTheBab** [comment]
```lsmod | grep vfio                                                                                               ✔  at 18:50:30 
vfio_pci               73728  0
vfio_iommu_type1       45056  0
vfio                   45056  2 vfio_iommu_type1,vfio_pci
vfio_virqfd            16384  1 vfio_pci
irqbypass              16384  2 vfio_pci,kvm
```

**Comment by GabTheBab** [comment]
```initrd=\booster-linux-zen.img root=LABEL=btrfsroot rw add_efi_memmap loglevel=3 quiet amd_iommu=on iommu=pt pcie_acs_override=downstream,multifunction vfio-pci.ids=10de:1b80,10de:10f0 booster.debug=1
```

**Comment by anatol** [comment]
The list of loaded modules is correct. The command line parameter is also in place.

I quickly looked at the mkinitcpio and dracut and I do not see any special avoid `vfio` modules. Their code just says load these `vfio` modules and that's it.

As a last attempt to solve this problem I pushed one more update to `modules_force_load` branch that loads `modules_force_load`, **waits** until they are fully loaded and only then it moves forward with the rest of the modules discovery. Please try it and see if it makes any difference.

**Comment by GabTheBab** [comment]
Sadly it still is loading the nvidia driver on commit 52144a5.

**Comment by anatol** [comment]
I merged `modules_force_load` config option to `master` in case if anyone wants to try to reproduce the issue.

**Comment by anatol** [comment]
I just added support for saving booster logs into `kmsg`/`journalctl`. It is available in `master` now.

It worth enabling this feature to get more information about what is going on at the boot time.

**Comment by GabTheBab** [comment]
```Mar 28 09:23:00 archlinux booster: no match found for alias acpi:LNXCPU:
Mar 28 09:23:00 archlinux booster: loading module vfio
Mar 28 09:23:00 archlinux booster: loading module vfio_virqfd
Mar 28 09:23:00 archlinux booster: loading module irqbypass
Mar 28 09:23:00 archlinux booster: no match found for alias acpi:LNXCPU:
Mar 28 09:23:00 archlinux booster: no match found for alias acpi:LNXCPU:
Mar 28 09:23:00 archlinux booster: no match found for alias acpi:LNXCPU:
Mar 28 09:23:00 archlinux booster: no match found for alias acpi:LNXCPU:
Mar 28 09:23:00 archlinux booster: no match found for alias acpi:LNXCPU:
Mar 28 09:23:00 archlinux booster: no match found for alias acpi:LNXCPU:
Mar 28 09:23:00 archlinux kernel: VFIO - User Level meta-driver version: 0.3
``` The nvidia driver loads after this, not before it, not sure why its grabbing onto it

**Comment by anatol** [comment]
Could you please make sure to add `printk.devkmsg=on` to disable kmsg throttling and then share full logs (at least the pats related to booster & vfio)?

> The nvidia driver loads after this, not before it, not sure why its grabbing onto it

From what I read above nvidia driver should not be loaded at all. But here you say you expect nvidia is loaded before vfio? What is the expected behavior?

**Comment by GabTheBab** [comment]
The expected behavior is the nvidia driver to not load, however the nvidia driver ends up loading after the messages I sent above.

**Comment by GabTheBab** [comment]
[log.txt](https://github.com/anatol/booster/files/6217616/log.txt)

**Comment by anatol** [comment]
@GabTheBab per your log.txt booster boots fine, loads vfio drivers, switches to the host userspace. nvidia driver is loaded in the userspace.

I compared your vfio output with this one https://wiki.archlinux.org/index.php/PCI_passthrough_via_OVMF#Verifying_that_the_configuration_worked and do not see following logs

```
[    0.341372] vfio_pci: add [10de:13c2[ffff:ffff]] class 0x000000/00000000
[    0.354704] vfio_pci: add [10de:0fbb[ffff:ffff]] class 0x000000/00000000
```

After reading kmod code I found that kernel does not use cmdline params for loadable modules. It was surprising for me. Instead it is expected that userspace handles cmdline parsing and provides required module params explicitly.

Here is a quick check to verify this hypothesis. @GabTheBab could you please compile and try following branch to see if it makes any difference for you. And if it does not please provide the dmesg logs. https://github.com/anatol/booster/tree/issue-20-quick-check

**Comment by GabTheBab** [comment]
That looks to have done it, vfio-pci is loaded and attached to the gpu.

**Comment by anatol** [comment]
Yay! It is great to hear the problem is solved.

I will work on #74 and provide a clean solution for your issue.

**Comment by GabTheBab** [comment]
Thanks for working on this issue, looking forward to the final solution in master

**Comment by anatol** [comment]
Done. It is part of `master` now. Please try it.

**Comment by GabTheBab** [comment]
Working on master now

**Comment by anatol** [comment]
This feature got released as part of booster-0.4 now in the stable repository.

I updated the wiki page with booster instructions https://wiki.archlinux.org/index.php/PCI_passthrough_via_OVMF#booster
But if you see other ways to improve the documentation please update it as well.
